### DISCUSSION TITLE
Start execution at a specific line in gcode

### DISCUSSION THREAD

**Discussion by dasarne** [terjeio/ioSender#227] [Ideas]
Would it be possible to **start from a certain line in the gcode**? If there are problems with the gcode or the machine, this can save the workpiece.

**Comment by phil-barrett** [comment]
This has been discussed. It is  very hard to do automatically. There is an Edit file feature that you can use to create a version of the loaded file that does just that. 

Note that there are a number of lines at the beginning of a GCode program that you should not delete.  They set up your workspace, turn on the spindle, set speed and so on. Also, you will not want to start at the last line executed but rather at some point earlier where there is a "full move" - one that fully specifies all your axis positions.

**Reply by dasarne** [reply]
Hi Phil,
I didn't want to do that automatically. It would be super cool if I could right click on a line in the source view of the gcode and then execution would start (or continue) from there. 
That would have many advantages: 
- I could just skip gcode that doesn't work for me at the moment
- I could repeat gcode
- I could extract something from a gcode that I need at that moment.
Speaking of daydreams: 

How cool would it be if I could set a breakpoint in gcode with the right mouse button where execution will just stop. 
Oh, and if I could then attach a macro to such a breakpoint, that would be heaven! ;-)

I think my wishes are based on a completely different way of working than isSender supports so far. 
But if you don't think about automatisms at all, this is quickly implemented and would be a great assistance.

I'm curious what you think of these ideas. 

Kind regards
 Arne

**Reply by terjeio** [reply]
Take a look at [this video](https://www.youtube.com/watch?v=ePlyXeXYZBE&list=PL1FfK7Yod27cy-MIhXsKkDwTmvclf7ps7&index=64) - it is not as simple as you may think...

> this is quickly implemented and would be a great assistance

You can do it? I have a large backlog to clear before I can attempt an implementation..

**Reply by terjeio** [reply]
Simple restart from block was implemented a while back, as were some other options such as copying a block to the MDI and selecting a number of blocks and only send them to the controller. Right click in the program listing for a popup menu with the options.
Restarting from a block and ensuring that the controller is at the correct position and in the correct state is not easy and is _not_ yet supported.

**Reply by terjeio** [reply]
> How cool would it be if I could set a breakpoint in gcode with the right mouse button where execution will just stop.

I have added this to the popup menu - it sends an `M0` _after_ the selected line has completed. The reason for this is since I have added a new  block marker `@` for the currently executing _motion_ line it was more logical to me that this should show up on the line where the break is inserted. Comments?

<img width="747" height="300" alt="image" src="https://github.com/user-attachments/assets/f32cc270-035c-4869-8fec-9999d6d7b0cd" />
<br><br>
Grbl, grblHAL and likely all other ports will only inform the currently executing line for lines containing motion, and only if line numbers are added to the real time report. This means that displaying the new block marker requires line numbers in the gcode program - which poses a new problem, not many has these days.
To get around this the sender will add line numbers when not present, if line numbers are enabled for the real time report (a $10 option flag). Line interval is 10. Is this ok, or should adding line numbers be a sender option instead of depending on a controller option?

The main purpose of new marker is to flag where to restart from if a job has been interrupted.
Here is an example, `ok` means the block has been acknowledged by the parser, `@` is the currently executing line and `*` is for blocks that has been buffered in the controller.

<img width="751" height="322" alt="image" src="https://github.com/user-attachments/assets/c0add187-46bc-454e-88a4-1bfeb429a80f" />
<br><br>
I have not pushed this to an edge version yet, more testing is needed. I may also try to change the real-time line number reporting to include non-motion line numbers (in grblHAL).
